src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/PackageSpecificWarningProperties.cs

```diff
@@ -36,22 +38,51 @@ public static PackageSpecificWarningProperties CreatePackageSpecificWarningPrope
             {
                 foreach (var framework in packageSpec.TargetFrameworks)
                 {
-                    warningProperties.AddRange(dependency.NoWarn, dependency.Name, framework.FrameworkName);
+                    warningProperties.AddRangeOfCodes(dependency.NoWarn, dependency.Name, framework.FrameworkName);
                 }
             }
 
             foreach (var framework in packageSpec.TargetFrameworks)
             {
                 foreach (var dependency in framework.Dependencies)
                 {
-                    warningProperties.AddRange(dependency.NoWarn, dependency.Name, framework.FrameworkName);
+                    warningProperties.AddRangeOfCodes(dependency.NoWarn, dependency.Name, framework.FrameworkName);
                 }
             }
 
             return warningProperties;
         }
 
         /// <summary>
+        /// Extracts PackageSpecific WarningProperties from a PackageSpec for a specific NuGetFramework
+        /// </summary>
+        /// <param name="packageSpec">PackageSpec containing the Dependencies with WarningProperties</param>
+        /// <param name="framework">NuGetFramework for which the properties should be assessed.</param>
+        /// <returns>PackageSpecific WarningProperties extracted from a PackageSpec for a specific NuGetFramework</returns>
+        public static PackageSpecificWarningProperties CreatePackageSpecificWarningProperties(PackageSpec packageSpec, 
+            NuGetFramework framework)
+        {
+            // NuGetLogCode -> LibraryId -> Set of Frameworks.
+            var warningProperties = new PackageSpecificWarningProperties();
+
+            foreach (var dependency in packageSpec.Dependencies)
+            {
+                warningProperties.AddRangeOfCodes(dependency.NoWarn, dependency.Name, framework);
+            }
+
+            var targetFrameworkInformation = packageSpec
+                .TargetFrameworks
+                .First(tfi => tfi.FrameworkName == framework);
```

Comment:
Use packageSpec.GetTargetFramework, it should have this optimization already, and it will ensure that if a comparer is needed that we already have it